wandb/sdk/launch/runner/aws.py

```diff
@@ -0,0 +1,475 @@
+                )
+                return None
+        _logger.info("Connecting to sagemaker client")
+
+        sagemaker_client = boto3.client(
+            "sagemaker",
+            region_name=region,
+            aws_access_key_id=access_key,
+            aws_secret_access_key=secret_key,
+        )
+        wandb.termlog(
+            "Launching run on sagemaker with entrypoint: {}".format(
+                " ".join(command_args)
+            )
+        )
+
+        sagemaker_args = build_sagemaker_args(launch_project, account_id, aws_tag)
+        _logger.info(f"Launching sagemaker job with args: {sagemaker_args}")
+        run = launch_sagemaker_job(launch_project, sagemaker_args, sagemaker_client)
+        if self.backend_config[PROJECT_SYNCHRONOUS]:
+            run.wait()
+        return run
+
+
+def aws_ecr_login(region: str, registry: str) -> str:
+    pw_command = f"aws ecr get-login-password --region {region}".split(" ")
+    try:
+        pw_process = subprocess.run(
+            pw_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True
+        )
+    except subprocess.CalledProcessError:
+        raise LaunchError(
+            "Unable to get login password. Please ensure you have AWS credentials configured"
+        )
+
+    try:
+        login_process = subprocess.run(
+            f"docker login --username AWS --password-stdin {registry}".split(" "),
+            input=pw_process.stdout,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            check=True,
+        )
+    except subprocess.CalledProcessError:
+        raise LaunchError("Failed to login to ECR")
+    return login_process.stdout.decode("utf-8")
+
+
+def merge_aws_tag_with_algorithm_specification(
+    algorithm_specification: Optional[Dict[str, Any]], aws_tag: Optional[str]
+) -> Dict[str, Any]:
+    """
+    AWS Sagemaker algorithms require a training image and an input mode.
+    If the user does not specify the specification themselves, define the spec
+    minimally using these two fields. Otherwise, if they specify the AlgorithmSpecification
+    set the training image if it is not set.
+    """
+    if algorithm_specification is None:
+        return {
+            "TrainingImage": aws_tag,
+            "TrainingInputMode": "File",
+        }
+    elif algorithm_specification.get("TrainingImage") is None:
+        algorithm_specification["TrainingImage"] = aws_tag
+    if algorithm_specification["TrainingImage"] is None:
+        raise LaunchError("Failed determine tag for training image")
+    return algorithm_specification
+
+
+def build_sagemaker_args(
+    launch_project: LaunchProject, account_id: str, aws_tag: Optional[str] = None,
+) -> Dict[str, Any]:
+    sagemaker_args = {}
+    resource_args = launch_project.resource_args
+    sagemaker_args["TrainingJobName"] = (
+        resource_args.get("TrainingJobName") or launch_project.run_id
+    )
+
+    sagemaker_args[
```

Comment:
it would be super cool if we maintained a directory somewhere (maybe in tests? idk) of sample/template `config.json` files for launch for each runner — you probably have one already for testing, adding that somewhere would be great

Reply:
I can add one to our tests directory.